test/tools/runner/hooks/leak_checker.ts

```diff
@@ -0,0 +1,182 @@
+      leakChecker.clients.add(this);
+      this[LeakChecker.kConnectCount] ??= 0;
+
+      const lastArg = args[args.length - 1];
+      const lastArgIsCallback = typeof lastArg === 'function';
+      if (lastArgIsCallback) {
+        const argsWithoutCallback = args.slice(0, args.length - 1);
+        return LeakChecker.originalConnect.call(this, ...argsWithoutCallback, (error, client) => {
+          if (error == null) {
+            this[LeakChecker.kConnectCount] += 1; // only increment on successful connects
+          }
+          return lastArg(error, client);
+        });
+      } else {
+        return LeakChecker.originalConnect.call(this, ...args).then(client => {
+          this[LeakChecker.kConnectCount] += 1; // only increment on successful connects
+          return client;
+        });
+      }
+    };
+
+    MongoClient.prototype.close = function (...args) {
+      this[LeakChecker.kConnectCount] ??= 0; // prevents NaN, its fine to call close on a client that never called connect
+      this[LeakChecker.kConnectCount] -= 1;
+      return LeakChecker.originalClose.call(this, ...args);
+    };
+  }
+
+  setup() {
+    this.setupSessionLeakChecker();
+    this.setupClientLeakChecker();
+  }
+
+  reset() {
+    for (const sessionPool of this.sessionPools) {
+      delete sessionPool[LeakChecker.kAcquiredCount];
+    }
+    ServerSessionPool.prototype.acquire = LeakChecker.originalAcquire;
+    ServerSessionPool.prototype.release = LeakChecker.originalRelease;
+    this.sessionPools.clear();
+
+    for (const client of this.clients) {
+      delete client[LeakChecker.kConnectCount];
+    }
+    MongoClient.prototype.connect = LeakChecker.originalConnect;
+    MongoClient.prototype.close = LeakChecker.originalClose;
+    this.clients.clear();
+  }
+
+  assert() {
+    for (const pool of this.sessionPools) {
+      expect(pool[LeakChecker.kAcquiredCount], 'ServerSessionPool acquired count').to.equal(0);
+    }
+    for (const client of this.clients) {
+      expect(client[LeakChecker.kConnectCount], 'MongoClient connect count').to.be.lessThanOrEqual(
+        0
+      );
+    }
+  }
+}
+
+let currentLeakChecker: LeakChecker | null;
+
+const leakCheckerBeforeEach = async function () {
+  currentLeakChecker = new LeakChecker(this.currentTest.fullTitle());
+  currentLeakChecker.setup();
+};
+
+const leakCheckerAfterEach = async function () {
+  let thrownError: Error | undefined;
+  try {
+    currentLeakChecker.assert();
+  } catch (error) {
+    thrownError = error;
+  }
+
+  currentLeakChecker?.reset();
+  currentLeakChecker = null;
+
+  if (thrownError instanceof Error) {
+    this.test.error(thrownError);
+  }
+};
+
+const TRACE_SOCKETS = process.env.TRACE_SOCKETS === 'true' ? true : false;
+const kSocketId = Symbol('socketId');
+const originalCreateConnection = net.createConnection;
+let socketCounter = 0n;
+
+const socketLeakCheckBeforeAll = function socketLeakCheckBeforeAll() {
```

Comment:
You don't use `socketCounter` anywhere to check for leaks.  Is the intention to use it eventually or did you mean to remove this?

Reply:
It's not meant to go up and down like the other counters, here it's acting as a unique id for each socket created, it helps you see when its logging if a socket that remains open is left over from a previous test